pkg/xds/server/callbacks/dataplane_lifecycle.go

```diff
@@ -155,33 +166,27 @@ func (d *DataplaneLifecycle) deregister(
 	log.Info("waiting for deregister proxy", "waitFor", d.deregistrationDelay)
 	<-time.After(d.deregistrationDelay)
 
-	d.lockMap.Lock(proxyKey)
-	defer d.lockMap.Unlock(proxyKey)
+	info.mtx.Lock()
+	defer info.mtx.Unlock()
 
-	// lock on the rest of the method to execute operations atomically:
-	// 1. check if proxy with the same 'proxyKey' connected to this or other CP instance
-	// 2. remove proxy resource from the store
-
-	_, ok := d.proxyTypeMap.Load(proxyKey)
-	if ok {
+	if info.connected {
 		log.Info("no need to deregister proxy. It has already connected to this instance")
 		return
 	}
 
-	connected, err := d.proxyConnectedToAnotherCP(ctx, proxyType, proxyKey, log)
-	if err != nil {
+	if connected, err := d.proxyConnectedToAnotherCP(ctx, proxyType, proxyKey, log); err != nil {
 		log.Error(err, "could not check if proxy connected to another CP")
 		return
-	}
-
-	if connected {
+	} else if connected {
 		return
 	}
 
 	log.Info("deregister proxy")
```

Comment:
Should you also keep a bool to indicate successful deletion to avoid double delete? (which could be fine thouhg)

Reply:
You're right, that's actually can cause a problem if ProxyConnect and ProxyReconnected are happening at pretty much the same time:

```
OnProxyConnected      │    OnProxyReconnected
──────────────────────┼──────────────────────────
                      │
LoadAndStore(:115)    │
                      │
info.mtx.Lock()(:119) │
                      │
Upsert(:124) with err │    LoadAndStore(:115)
                      │
Delete(:133)          │
                      │
info.mtx.Unlock()     │
                      │
                      │    info.mtx.Lock()(:119)
                      │
                      │    Upsert(:124) no err
                      │
                      │    info.connected = true(:138)
                      │
                      │    info.mtx.Unlock()
──────────────────────┴───────────────────────────
```
So we have the resource in the store, but not in the map. I guess `deleted bool` flag in proxyInfo is going to fix it.

Reply:
TBH I now I'm more convinced that the 2-map approach was better :)

Reply:
Haha up to you really.

Reply:
Ended up adding `deleted bool` flag in proxyInfo